File: src/java/magma/option/None.java
package magma.option;

import java.util.function.Consumer;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.function.Supplier;

public class None<T> implements Option<T> {
    @Override
    public <R> Option<R> map(Function<T, R> mapper) {
        return new None<>();
    }

    @Override
    public T orElseGet(Supplier<T> other) {
        return other.get();
    }

    @Override
    public Tuple<Boolean, T> toTuple(T other) {
        return new Tuple<>(false, other);
    }

    @Override
    public void ifPresent(Consumer<T> consumer) {
    }

    @Override
    public T orElse(T other) {
        return other;
    }

    @Override
    public Option<T> filter(Predicate<T> predicate) {
        return new None<>();
    }

    @Override
    public boolean isPresent() {
        return false;
    }
}

File: src/clang/magma/option/None.c
#include "None.h"
magma.option.Option<magma.option.R> map(magma.option.R(*mapper)(T)){return ();
}
T orElseGet(T(*other)()){return other.get();
}
magma.option.Tuple<magma.option.Boolean, T> toTuple(T other){return (false, other);
}
magma.option.void ifPresent(void(*consumer)(T)){
}
T orElse(T other){return other;
}
magma.option.Option<T> filter(int(*predicate)(T)){return ();
}
magma.option.boolean isPresent(){return false;
}
magma.option.R match(magma.option.R(*ifPresent)(T), magma.option.R(*ifEmpty)()){return ifEmpty.get();
}
magma.option.boolean isEmpty(){return true;
}
magma.option.Option<T> or(magma.option.Option<T>(*other)()){return other.get();
}
magma.option.Option<magma.option.R> flatMap(magma.option.Option<magma.option.R>(*mapper)(T)){return ();
}

